dimod/include/dimod/abc.h

```diff
@@ -0,0 +1,1083 @@
+        asym_quadratic(v, u) += bias;
+    }
+}
+
+template <class bias_type, class index_type>
+void QuadraticModelBase<bias_type, index_type>::add_quadratic(
+        std::initializer_list<index_type> row, std::initializer_list<index_type> col,
+        std::initializer_list<bias_type> biases) {
+    auto rit = row.begin();
+    auto cit = col.begin();
+    auto bit = biases.begin();
+    for (; rit < row.end() && cit < col.end() && bit < biases.end(); ++rit, ++cit, ++bit) {
+        add_quadratic(*rit, *cit, *bit);
+    }
+}
+
+template <class bias_type, class index_type>
+template <class ItRow, class ItCol, class ItBias>
+void QuadraticModelBase<bias_type, index_type>::add_quadratic(ItRow row_iterator,
+                                                              ItCol col_iterator,
+                                                              ItBias bias_iterator,
+                                                              index_type length) {
+    // todo: performance testing of adding and then sorting later
+    for (index_type i = 0; i < length; ++i, ++row_iterator, ++col_iterator, ++bias_iterator) {
+        add_quadratic(*row_iterator, *col_iterator, *bias_iterator);
+    }
+}
+
+template <class bias_type, class index_type>
+void QuadraticModelBase<bias_type, index_type>::add_quadratic_back(index_type u, index_type v,
+                                                                   bias_type bias) {
+    assert(0 <= u && static_cast<size_t>(u) <= num_variables());
+    assert(0 <= v && static_cast<size_t>(v) <= num_variables());
+
+    enforce_adj();
+
+    // check the condition for adding at the back
+    assert((*adj_ptr_)[v].empty() || (*adj_ptr_)[v].back().v <= u);
+    assert((*adj_ptr_)[u].empty() || (*adj_ptr_)[u].back().v <= v);
+
+    if (u == v) {
+        switch (this->vartype(u)) {
+            case Vartype::BINARY: {
+                // 1*1 == 1 and 0*0 == 0 so this is linear
+                add_linear(u, bias);
+                break;
+            }
+            case Vartype::SPIN: {
+                // -1*-1 == +1*+1 == 1 so this is a constant offset
+                offset_ += bias;
+                break;
+            }
+            default: {
+                // self-loop
+                (*adj_ptr_)[u].emplace_back(v, bias);
+                break;
+            }
+        }
+    } else {
+        (*adj_ptr_)[u].emplace_back(v, bias);
+        (*adj_ptr_)[v].emplace_back(u, bias);
+    }
+}
+
+template <class bias_type, class index_type>
+template <class T>
+void QuadraticModelBase<bias_type, index_type>::add_quadratic_from_dense(const T dense[],
+                                                                         index_type num_variables) {
+    static_assert(std::is_arithmetic<T>::value, "T must be numeric");
+    assert(0 <= num_variables);
+    assert(static_cast<size_type>(num_variables) <= this->num_variables());
+
+    enforce_adj();
+
+    if (is_linear()) {
```

Comment:
This code could be made faster with some fancy tricks, but I guess that is not too important.

Reply:
Agree, but let's leave it alone for right now.